File: mockserver/mockserver-client-java/src/main/java/org/mockserver/client/TurnBuilder.java
package org.mockserver.client;

import org.mockserver.llm.ParsedMessage;
import org.mockserver.mock.Expectation;
import org.mockserver.model.Completion;
import org.mockserver.model.NormalizationOptions;

import java.util.regex.Pattern;

/**
 * Sub-builder for configuring a single turn within an LLM conversation mock.
 * <p>
 * Provides fluent predicates for matching inbound requests and a response
 * to return when all predicates match. Returns to the parent
 * {@link LlmConversationBuilder} via {@link #andThen()}.
 */
public class TurnBuilder {

    private final LlmConversationBuilder parent;
    Integer turnIndex;
    String latestMessageContains;
    Pattern latestMessageMatches;
    ParsedMessage.Role latestMessageRole;
    String containsToolResultFor;
    NormalizationOptions normalization;
    Completion completion;

    TurnBuilder(LlmConversationBuilder parent) {
        this.parent = parent;
    }

    /**
     * Match when the conversation has exactly {@code n} assistant turns.
     *
     * @param n the turn index (0-based count of assistant messages)
     * @return this builder
     */
    public TurnBuilder whenTurnIndex(int n) {
        this.turnIndex = n;
        return this;
    }

    /**
     * Match when the latest message's text content contains the given substring.
     *
     * @param text the substring to match
     * @return this builder
     */
    public TurnBuilder whenLatestMessageContains(String text) {
        this.latestMessageContains = text;
        return this;
    }

    /**
     * Match when the latest message's text content matches the given regex.
     *
     * @param regex the regex pattern to match
     * @return this builder
     * @throws IllegalArgumentException if the regex fails to compile
     */
    public TurnBuilder whenLatestMessageMatches(Pattern regex) {
        if (regex == null) {
            throw new IllegalArgumentException("regex must not be null");
        }
        this.latestMessageMatches = regex;
        return this;
    }

    /**
     * @deprecated misleading name — this overload matches by regex, not substring.
     * Use {@link #whenLatestMessageMatches(Pattern)}. Retained for source
     * compatibility; delegates to {@code whenLatestMessageMatches}.
     */
    @Deprecated
    public TurnBuilder whenLatestMessageContains(Pattern regex) {
        return whenLatestMessageMatches(regex);
    }

    /**
     * Match when the latest message has the given role.
     *
     * @param role the role to match
     * @return this builder
     */
    public TurnBuilder whenLatestMessageRole(ParsedMessage.Role role) {
        this.latestMessageRole = role;
        return this;
    }

    /**
     * Match when the conversation contains a tool result for a prior call
     * of the named tool.
     *
     * @param toolName the tool name to look for
     * @return this builder
     */
    public TurnBuilder whenContainsToolResultFor(String toolName) {
        this.containsToolResultFor = toolName;
        return this;
    }

    /**
     * Apply opt-in prompt normalisation before the {@code whenLatestMessage…}
     * text predicates are evaluated, so cosmetic differences (whitespace, JSON
     * key ordering, volatile ids/timestamps) do not block a match. Deterministic
     * — the same prompt always normalises identically.
     *
     * @param normalization the normalisation options
     * @return this builder
     */
    public TurnBuilder withNormalization(NormalizationOptions normalization) {
        this.normalization = normalization;
        return this;
    }

    /**
     * Set the completion to return for this turn.
     *
     * @param completion the completion response
     * @return this builder
     */
    public TurnBuilder respondingWith(Completion completion) {
        this.completion = completion;
        return this;
    }

    /**
     * Return to the parent conversation builder for chaining additional turns.
     *
     * @return the parent LlmConversationBuilder
     */
    public LlmConversationBuilder andThen() {
        return parent;
    }

    /**
     * Shortcut: build and register all turns with the MockServerClient.
     * Delegates to the parent builder's {@code applyTo}.
     *
     * @param client the MockServerClient
     * @return the created expectations
     */
    public Expectation[] applyTo(MockServerClient client) {
        return parent.applyTo(client);
    }
}

File: mockserver-client-php/src/Llm/TurnBuilder.php
<?php

declare(strict_types=1);

namespace MockServer\Llm;

use InvalidArgumentException;
use MockServer\Expectation;
use MockServer\MockServerClient;

/**
 * Sub-builder configuring one turn of a conversation mock
 * (mirrors org.mockserver.client.TurnBuilder).
 */
class TurnBuilder
{
    private LlmConversationBuilder $parent;

    public ?int $turnIndex = null;
    public ?string $latestMessageContains = null;
    public ?string $latestMessageMatches = null;
    public ?string $latestMessageRole = null;
    public ?string $containsToolResultFor = null;
    public ?string $semanticMatchAgainst = null;
    public ?NormalizationOptions $normalization = null;
    /** @var array<string, mixed>|null */
    public ?array $chaos = null;
    public ?Completion $completion = null;

    public function __construct(LlmConversationBuilder $parent)
    {
        $this->parent = $parent;
    }

    public function whenTurnIndex(int $turnIndex): self
    {
        $this->turnIndex = $turnIndex;
        return $this;
    }

    public function whenLatestMessageContains(string $text): self
    {
        $this->latestMessageContains = $text;
        return $this;
    }

    public function whenLatestMessageMatches(string $regex): self
    {
        if ($regex === '') {
            throw new InvalidArgumentException('regex must not be empty');
        }
        $this->latestMessageMatches = $regex;
        return $this;
    }

    public function whenLatestMessageRole(string $role): self
    {
        $this->latestMessageRole = $role;
        return $this;
    }

    public function whenContainsToolResultFor(string $toolName): self
    {
        $this->containsToolResultFor = $toolName;
        return $this;
    }

    public function whenSemanticMatch(string $expectedMeaning): self
    {
        $this->semanticMatchAgainst = $expectedMeaning;
        return $this;
    }

    public function withNormalization(NormalizationOptions $normalization): self
    {
        $this->normalization = $normalization;
        return $this;
    }

    /**
     * Attach a chaos/fault profile (a JSON-serialisable map) to this turn.
     *
     * @param array<string, mixed> $chaos
     */
    public function withChaos(array $chaos): self
    {
        $this->chaos = $chaos;
        return $this;
    }

    public function respondingWith(Completion $completion): self
    {
        $this->completion = $completion;
        return $this;
    }

    /**
     * Start a new turn on the parent conversation.
     */
    public function turn(): TurnBuilder
    {
        return $this->parent->turn();
    }

    /**
     * Return to the parent conversation builder.
     */
    public function andThen(): LlmConversationBuilder
    {
        return $this->parent;
    }

    /**
     * @return list<Expectation>
     */
    public function build(): array
    {
        return $this->parent->build();
    }

    /**
     * @return array<mixed>
     */
    public function applyTo(MockServerClient $client): array
    {
        return $this->parent->applyTo($client);
    }

    /**
     * Build the conversation predicates for this turn (always populated; the
     * caller decides whether to attach it based on {@see hasAnyPredicate()}).
     */
    public function predicates(): ConversationPredicates
    {
        $predicates = new ConversationPredicates();
        if ($this->turnIndex !== null) {
            $predicates->withTurnIndex($this->turnIndex);
        }
        if ($this->latestMessageContains !== null) {
            $predicates->withLatestMessageContains($this->latestMessageContains);
        }
        if ($this->latestMessageMatches !== null) {
            $predicates->withLatestMessageMatches($this->latestMessageMatches);
        }
        if ($this->latestMessageRole !== null) {
            $predicates->withLatestMessageRole($this->latestMessageRole);
        }
        if ($this->containsToolResultFor !== null) {
            $predicates->withContainsToolResultFor($this->containsToolResultFor);
        }
        if ($this->semanticMatchAgainst !== null) {
            $predicates->withSemanticMatchAgainst($this->semanticMatchAgainst);
        }
        if ($this->normalization !== null) {
            $predicates->withNormalization($this->normalization);
        }
        return $predicates;
    }

    /**
     * True if at least one predicate (NOT normalization) is set.
     */
    public function hasAnyPredicate(): bool
    {
        return $this->turnIndex !== null
            || $this->latestMessageContains !== null
            || $this->latestMessageMatches !== null
            || $this->latestMessageRole !== null
            || $this->containsToolResultFor !== null
            || $this->semanticMatchAgainst !== null;
    }
}
